src/services/transaction_watcher_signer_service.ts

```diff
@@ -0,0 +1,364 @@
+import { RPCSubprovider, Web3ProviderEngine } from '@0x/subproviders';
+import { BigNumber, intervalUtils, providerUtils } from '@0x/utils';
+import { SupportedProvider, Web3Wrapper } from '@0x/web3-wrapper';
+import { utils as web3WrapperUtils } from '@0x/web3-wrapper/lib/src/utils';
+import { Connection, Not, Repository } from 'typeorm';
+
+import { ETHEREUM_RPC_URL, META_TXN_RELAY_PRIVATE_KEYS } from '../config';
+import {
+    ETH_GAS_STATION_API_BASE_URL,
+    EXPECTED_MINED_SEC,
+    NUMBER_OF_BLOCKS_UNTIL_CONFIRMED,
+    ONE_SECOND_MS,
+    TX_WATCHER_POLLING_INTERVAL_MS,
+    UNSTICKING_TRANSACTION_GAS_MULTIPLIER,
+} from '../constants';
+import { TransactionEntity } from '../entities';
+import { logger } from '../logger';
+import { TransactionStates } from '../types';
+import { Signer } from '../utils/signer';
+
+export class TransactionWatcherSignerService {
+    private readonly _transactionRepository: Repository<TransactionEntity>;
+    private readonly _provider: SupportedProvider;
+    private readonly _web3Wrapper: Web3Wrapper;
+    private readonly _transactionWatcherTimer: NodeJS.Timer;
+    private readonly _signers: Map<string, Signer>;
+    private readonly _availableSignerPublicAddresses: string[];
+
+    private static _createWeb3Provider(rpcHost: string): SupportedProvider {
+        const providerEngine = new Web3ProviderEngine();
+        providerEngine.addProvider(new RPCSubprovider(rpcHost));
+        providerUtils.startProviderEngine(providerEngine);
+        return providerEngine;
+    }
+    constructor(dbConnection: Connection) {
+        this._provider = TransactionWatcherSignerService._createWeb3Provider(ETHEREUM_RPC_URL);
+        this._transactionRepository = dbConnection.getRepository(TransactionEntity);
+        this._web3Wrapper = new Web3Wrapper(this._provider);
+        this._signers = new Map<string, Signer>();
+        this._availableSignerPublicAddresses = META_TXN_RELAY_PRIVATE_KEYS.map(key => {
+            const signer = new Signer(key, ETHEREUM_RPC_URL);
+            this._signers.set(signer.publicAddress, signer);
+            return signer.publicAddress;
+        });
+        this._transactionWatcherTimer = intervalUtils.setAsyncExcludingInterval(
+            async () => {
+                logger.trace('syncing transaction status');
+                await this.syncTransactionStatusAsync();
+            },
+            TX_WATCHER_POLLING_INTERVAL_MS,
+            (err: Error) => {
+                logger.error({
+                    message: `transaction watcher failed to sync transaction status: ${JSON.stringify(err)}`,
+                    err: err.stack,
+                });
+            },
+        );
+    }
+    public stop(): void {
+        intervalUtils.clearAsyncExcludingInterval(this._transactionWatcherTimer);
+    }
+    public async syncTransactionStatusAsync(): Promise<void> {
+        await this._signAndBroadcastTransactionsAsync();
```

Comment:
suggestion: the underlying can throw, which means we wouldn't continue to the next stage, consider making them not throw or try catching each one separately?

I.e if we can't sign and broadcast (out of money?), we still want to try syncing, checking for stuck and checking for confirmed.

Reply:
Definitely, added a catch here.